File: jobs/2026-05-30__07-57-35/cron_job_cleanup_ts__boHAwMx/artifacts/user/cron-app-zr-bohawmx/cleanup/cleanup.ts
import { SQLDatabase } from "encore.dev/storage/sqldb";
import { cron } from "encore.dev/cron";
import { api } from "encore.dev/api";

// Define the database
export const recordsDB = new SQLDatabase("records_db", {
  migrations: "./migrations",
});

// Define types
interface RecordItem {
  id: number;
  data: string;
  created_at: string;
}

interface CreateRecordRequest {
  data: string;
  created_at?: string;
}

interface CreateRecordResponse {
  id: number;
}

interface GetRecordsResponse {
  records: RecordItem[];
}

interface CleanupResponse {
  deleted_count: number;
}

// POST /records - Insert a new record
export const createRecord = api(
  { method: "POST", path: "/records", expose: true },
  async (req: CreateRecordRequest): Promise<CreateRecordResponse> => {
    const createdAt = req.created_at ?? new Date().toISOString();
    const result = await recordsDB.queryRow`
      INSERT INTO records (data, created_at) VALUES (${req.data}, ${createdAt}::timestamp) RETURNING id
    `;
    return { id: result!.id };
  }
);

// GET /records - Get all records ordered by id
export const getRecords = api(
  { method: "GET", path: "/records", expose: true },
  async (): Promise<GetRecordsResponse> => {
    const rows = await recordsDB.query`
      SELECT id, data, created_at FROM records ORDER BY id
    `;
    const records: RecordItem[] = [];
    for await (const row of rows) {
      records.push({
        id: row.id,
        data: row.data,
        created_at: row.created_at,
      });
    }
    return { records };
  }
);

// POST /cleanup - Delete records older than 1 hour
export const cleanup = api(
  { method: "POST", path: "/cleanup", expose: true },
  async (): Promise<CleanupResponse> => {
    const result = await recordsDB.queryRow`
      WITH deleted AS (
        DELETE FROM records WHERE created_at < NOW() - INTERVAL '1 hour' RETURNING id
      ) SELECT count(*)::int AS deleted_count FROM deleted
    `;
    return { deleted_count: result?.deleted_count ?? 0 };
  }
);

// Cron job that runs every 1 hour
const _cleanupJob = cron("cleanup-job", {
  every: "1h",
  endpoint: cleanup,
});

File: jobs/2026-07-03__05-39-37/encore_cron_job_cleanup_go__jie56oj/artifacts/user/cleanup-app/cleanup/cleanup.go
// Package cleanup implements a service that stores records in a PostgreSQL
// database and periodically cleans up stale records older than 24 hours.
package cleanup

import (
	"context"
	"fmt"
	"time"

	"encore.dev/cron"
	"encore.dev/storage/sqldb"
)

// cleanupDB is the PostgreSQL database used by the cleanup service.
var cleanupDB = sqldb.NewDatabase("cleanup_db", sqldb.DatabaseConfig{
	Migrations: "./migrations",
})

// Record represents a single record stored in the database.
type Record struct {
	ID        string    `json:"id"`
	Data      string    `json:"data"`
	CreatedAt time.Time `json:"created_at"`
}

// CreateRecordRequest is the request body accepted by the POST /records endpoint.
type CreateRecordRequest struct {
	ID        string `json:"id"`
	Data      string `json:"data"`
	CreatedAt string `json:"created_at"` // RFC3339 formatted timestamp
}

// ListRecordsResponse is the response returned by the GET /records endpoint.
type ListRecordsResponse struct {
	Records []Record `json:"records"`
}

// CreateRecord inserts a new record into the database. The created_at field is
// provided in RFC3339 format so that historical records can be inserted for
// testing the cleanup logic.
//
//encore:api public method=POST path=/records
func CreateRecord(ctx context.Context, req *CreateRecordRequest) error {
	createdAt, err := time.Parse(time.RFC3339, req.CreatedAt)
	if err != nil {
		return fmt.Errorf("invalid created_at (expected RFC3339): %w", err)
	}
	_, err = cleanupDB.Exec(ctx, `
		INSERT INTO records (id, data, created_at)
		VALUES ($1, $2, $3)
	`, req.ID, req.Data, createdAt)
	if err != nil {
		return fmt.Errorf("insert record: %w", err)
	}
	return nil
}

// ListRecords returns every record currently stored in the database.
//
//encore:api public method=GET path=/records
func ListRecords(ctx context.Context) (*ListRecordsResponse, error) {
	rows, err := cleanupDB.Query(ctx, `
		SELECT id, data, created_at FROM records ORDER BY created_at
	`)
	if err != nil {
		return nil, fmt.Errorf("query records: %w", err)
	}
	defer rows.Close()

	records := []Record{}
	for rows.Next() {
		var r Record
		if err := rows.Scan(&r.ID, &r.Data, &r.CreatedAt); err != nil {
			return nil, fmt.Errorf("scan record: %w", err)
		}
		records = append(records, r)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate records: %w", err)
	}
	return &ListRecordsResponse{Records: records}, nil
}

// Cleanup deletes all records whose created_at timestamp is older than 24
// hours from the current time. It is exposed as a public endpoint so that it
// can be triggered manually for verification, and is also invoked on a
// schedule by the cron job defined below.
//
//encore:api public method=POST path=/cleanup
func Cleanup(ctx context.Context) error {
	_, err := cleanupDB.Exec(ctx, `
		DELETE FROM records WHERE created_at < NOW() - INTERVAL '24 hours'
	`)
	if err != nil {
		return fmt.Errorf("cleanup records: %w", err)
	}
	return nil
}

// cleanupJob runs the Cleanup endpoint every hour to remove stale records.
var _ = cron.NewJob("cleanup-job", cron.JobConfig{
	Title:    "Clean up stale records",
	Every:    1 * cron.Hour,
	Endpoint: Cleanup,
})